Koha/REST/V1/Lists.pm

```diff
@@ -16,17 +16,323 @@ package Koha::REST::V1::Lists;
 # along with Koha; if not, see <http://www.gnu.org/licenses>.
 
 use Modern::Perl;
-
 use Mojo::Base 'Mojolicious::Controller';
-
+use Koha::Virtualshelf;
 use Koha::Virtualshelves;
-
-use Try::Tiny qw( catch try );
+use Koha::Virtualshelfcontent;
 
 =head1 API
 
 =head2 Methods
 
+=head3 list
+
+List all virtual shelves
+
+=cut
+
+sub list {
+    my $c = shift->openapi->valid_input or return;
+
+    my @lists;
+    my $virtualshelves = Koha::Virtualshelves->search();
+    while (my $list = $virtualshelves->next) {
+        push @lists, {
+            list_id       => $list->id,
+            name => $list->shelfname,
+            creation_date => $list->created_on,
+            owner_id    => $list->owner,
+            allow_change_from_others => $list->allow_change_from_others,
+            allow_change_from_owner => $list->allow_change_from_owner,
+            public   => $list->public,
+            default_sort_field => $list->sortfield,
+            records  => \@records,
+        };
+    }
+
+    return $c->render(json => \@lists);
+}
+
+=head3 create
+
+Create a virtual shelf
+
+=cut
+
+sub create {
+    my $c = shift->openapi->valid_input or return;
+    my $params = $c->req->json;
+
+    my $list = Koha::Virtualshelf->new({
+        shelfname => $params->{shelfname},
+        owner     => $params->{owner},
+        public    => $params->{public} // 0,
+    })->store;
+
+    # Add biblios to the newly created list
+    if (exists $params->{biblionumbers} && ref $params->{biblionumbers} eq 'ARRAY') {
+        foreach my $biblionumber (@{$params->{biblionumbers}}) {
+            Koha::Virtualshelfcontent->new({
+                shelfnumber  => $list->shelfnumber,
+                biblionumber => $biblionumber,
+                dateadded    => DateTime->now,
+                borrowernumber => $params->{owner},
+            })->store;
+        }
+    }
+
+    return $c->render(status => 201, json => { id => $list->id });
+}
+
+=head3 read
```

Comment:
Please call this method 'get'.. 'read' is a homonym for a Perl builtin so can sometimes be confused... also, that would make this consistent with other controllers.